content/courses/Ada_For_The_C_Embedded_Developer/chapters/05_SPARK.rst

```diff
@@ -7,36 +7,105 @@ Enhancing Verification with SPARK and Ada
 
 .. include:: ../../global.txt
 
-Dealing or not Dealing with Exceptions
---------------------------------------
-
-In Ada, incorrect execution of code is often translated into what's called an
-exception, which interrupts the normal flow of execution. For example,
-accessing to an index out of the bound of an array will raise such an event.
-The general model for Ada exceptions is the same as of many other programming
-languages, where the exception is being propagated from call frame to call
-frame up until it either reaches the main subprogram (terminating the
-application) or is caught by an so called exception handler in one of the call
-frames. For example:
+Understanding Exceptions and Dynamic Checks
+-------------------------------------------
+
+In Ada, several common programming errors that are not already detected 
+at compile-time are detected instead at run-time, triggering 
+"exceptions" that interrupt the normal flow of execution. For example, 
+an exception is raised by an attempt to access an array component via an 
+index that is out of bounds, thereby precluding exploits based on buffer 
+overflow. Several other cases also raise language-defined exceptions, 
+such as scalar range constraint violations, and users may declare and 
+raise their own application-specific exceptions. 
+
+We will loosely define a "frame" as a language construct that has a call
+stack entry when called, for example a procedure or function body. There
+are a few other constructs that are also pertinent but this definition
+will suffice for now.
+
+Frames can have "exception handlers" that specify the response to be
+executed when an exception is "raised." These handlers can be specified to
+"handle" one or more specific exceptions, thus allowing the response to
```

Comment:
Editorial suggestion: although this sentence is correct, it might be too formal / abstract for the more casual reader.